src/Razor/src/Microsoft.AspNetCore.Razor.ExternalAccess.RoslynWorkspace/RazorProjectJsonSerializer.cs

```diff
@@ -0,0 +1,215 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System.Diagnostics.CodeAnalysis;
+using System.Runtime.InteropServices;
+using Microsoft.AspNetCore.Razor.Language;
+using Microsoft.AspNetCore.Razor.ProjectEngineHost;
+using Microsoft.AspNetCore.Razor.ProjectEngineHost.Serialization;
+using Microsoft.AspNetCore.Razor.Telemetry;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Razor;
+using Newtonsoft.Json;
+
+namespace Microsoft.AspNetCore.Razor.ExternalAccess.RoslynWorkspace;
+
+internal static class RazorProjectJsonSerializer
+{
+    private static readonly JsonSerializer s_serializer;
+    private static readonly EmptyProjectEngineFactory s_fallbackProjectEngineFactory;
+    private static readonly StringComparison s_stringComparison;
+
+    static RazorProjectJsonSerializer()
+    {
+        s_serializer = new JsonSerializer()
+        {
+            Formatting = Formatting.Indented
+        };
+
+        s_serializer.Converters.RegisterProjectSerializerConverters();
+
+        s_fallbackProjectEngineFactory = new EmptyProjectEngineFactory();
+        s_stringComparison = RuntimeInformation.IsOSPlatform(OSPlatform.Linux)
+            ? StringComparison.Ordinal
+            : StringComparison.OrdinalIgnoreCase;
+    }
+
+    public static async Task SerializeAsync(Project project, string projectRazorJsonFileName, CancellationToken cancellationToken)
```

Comment:
This is new code, essentially amalgamating code from various layers into the minimal "host the razor project engine" class possible.